jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java

```diff
@@ -927,6 +928,34 @@ default int size()
         return size;
     }
 
+    /**
+     * @param fields the {@link HttpFields} to convert to a {@link Map}.
+     * @return an unmodifiable {@link Map} representing the contents of the {@link HttpFields}.
+     */
+    static Map<String, List<String>> asMap(HttpFields fields)
+    {
+        Map<String, List<String>> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+        for (HttpField f : fields)
+        {
+            if (!headers.containsKey(f.getName()))
+            {
+                HttpHeader header = f.getHeader();
+                List<String> values = header == null ? fields.getValuesList(f.getName()) : fields.getValuesList(header);
+                headers.put(f.getName(), Collections.unmodifiableList(values));
+            }
+        }
+        return Collections.unmodifiableMap(headers);
+    }
+
+    /**
+     * @param fields the {@link HttpFields} to convert to a {@link Map}.
+     * @return a {@link Map} where changes to the contents will be reflected in the supplied {@link HttpFields}.
+     */
+    static Map<String, List<String>> asMap(HttpFields.Mutable fields)
+    {
+        return new HttpFieldsMap(fields);
+    }
```

Comment:
I'm not sure that using a separate signature is the right way to go, as often a mutable HttpFields is not held as a Mutable.    I think an instanceof Mutable check in the other method is a better way to go.

Reply:
I think I might just separate these into separate methods, `asMap()` and `asMutableMap()`, because I think its clearer reading the code to understand what the intention is. And you can check all the places each of them are called to verify they are correct.